atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java

```diff
@@ -565,14 +566,36 @@ private RemoteClientConnection getOrCreateClientConnection(final Channel channel
     bootstrap.channel(clientChannelClass);
     bootstrap.remoteAddress(resolvedAddress, address.port());
     bootstrap.handler(new BasicClientChannelInitializer(future));
-    bootstrap
-        .connect()
+    final Channel channel =
+        bootstrap
+            .connect()
+            .addListener(
+                f -> {
+                  if (!f.isSuccess()) {
+                    future.completeExceptionally(f.cause());
+                  }
+                })
+            .channel();
+
+    // immediately ensure we're notified of the channel being closed. the common case is that the
+    // future is already completed by the time the channel is closed, so this really only to ensure
+    // that the future is completed even if the channel is closed before we manually complete (i.e.
+    // post handshake)
+    channel
+        .closeFuture()
         .addListener(
-            f -> {
-              if (!f.isSuccess()) {
-                future.completeExceptionally(f.cause());
+            onClose -> {
+              if (onClose.isSuccess()) {
+                future.completeExceptionally(
+                    new ConnectException(
+                        String.format(
+                            "Channel %s for address %s was closed before the request could be completed",
+                            channel, address)));
+              } else {
+                future.completeExceptionally(onClose.cause());
               }
```

Comment:
Sorry I'm not so into this API. This addition is not so obvious to me sorry :see_no_evil: 

I assume the listener for close will be called also when our request went trough? How can we detect that the request wasn't sent? What does `onClose.isSuccess` even mean? Why we have no message on the else branch? Shouldn't we check whether our future is not yet completed? Or is it fine the thrown an exception in this listener?

Reply:
The future is called when it's completed AFAIK. We don't need to detect that the request was or wasn't sent - it's a bit tricky because it relies on the order of callbacks and when they're called. There is one edge case I suppose where the channel may get closed while we're waiting for the response. I can reword the message so it's more about the channel was closed before the request was processed correctly. Or is that still not clear?

So normally the channel is closed at the end, right, when we're done with it. At that point though, we've handled the request, and we've already either completed the future successfully or with an exception. So in the common case, where we close the channel at the end when we're done, nothing really happens here since the future is already completed.

There was an edge case that we noticed in the AdvertisedAddressTest where the channel was closed mid-way through the handshake - it was closed successfully mind, and I never figured out why or who was closing it. At any rate, from a `sendAndReceive` perspective (i.e. contract), this is a failure. We didn't send nor receive the request, so we shouldn't mark it as successfully completed. If the channel is closed _before_ anything else completes the future, imo, it's a breach of contract and clearly it's a failure. If the channel was closed with an error, then we can report that error. If it wasn't, the best we can do is that something closed it before we could fully handle it (but unfortunately it's difficult to know what).

I hope that clears things up. To be honest, this whole API is a overly complex I think. I don't think Netty is per se, but how the transport here is implemented makes things much more complex than they need to be. I can't wait to switch to gRPC.